url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/primitives/CompressedString.java

```diff
@@ -0,0 +1,205 @@
+/**
+ * This file is distributed under a BSD-style license. See the included LICENSE.txt file
+ * for more information. 
+ * Copyright (c) 2009, James Halliday
+ * All rights reserved.
+ */
+package malictus.klang.primitives;
+
+import java.io.*;
+import java.nio.*;
+import java.nio.charset.*;
+import malictus.klang.*;
+
+/**
+ * Primitive representing a plain, unformatted string compressed using the ZLIB library.
+ * @author Jim Halliday
+ */
+public class CompressedString implements Primitive, StringPrim {
+	
+	private String value;
+	private String textEncoding;
+	private boolean canChangeEncoding;
+	
+	/**
+	 * Constructor from a file reference. The value will be read from the file at
+	 * the current file pointer, and the file pointer will be advanced to 
+	 * just beyond the compressed string.
+	 * 
+	 * @param raf a RandomAccessFile object that points to a valid file, queued for reading
+	 * @param stoppoint the byte position to stop reading data, or -1 if entire file should be read
+	 * @param textEncoding the text encoding for this string
+	 * @param canChangeEncoding true if text encoding can change, and false otherwise
+	 * @throws IOException if the file cannot be read properly
+	 * @throws BadValueException if value is incorrect
+	 * @throws CharacterCodingException if a character encoding problem occurs
+	 */
+	public CompressedString(RandomAccessFile raf, long stoppoint, String textEncoding, boolean canChangeEncoding) throws IOException, BadValueException, CharacterCodingException {
+		this.textEncoding = textEncoding;
+		this.canChangeEncoding = canChangeEncoding;
+		if (stoppoint != -1) {
+			if (raf.getFilePointer() >= stoppoint) {
+				value = "";
+				return;
+			}
+		}
+		//check for pad byte at end
+		long pos = raf.getFilePointer();
+		raf.seek(stoppoint - 1);
+		byte x = raf.readByte();
+		if (x == 0) {
+			stoppoint = stoppoint - 1;
+		}
+		raf.seek(pos);
+		byte[] byt = new byte[(int)(stoppoint - raf.getFilePointer())];
+		raf.read(byt);
+		byte[] inflated = KlangUtil.inflate(byt);
+		try {
+			value = new String(inflated, 0, inflated.length, textEncoding);
+		} catch (Exception err) {
+			err.printStackTrace();
+			throw new BadValueException();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 57:62, 118:123.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.